02-peer-protocol.md

```diff
@@ -49,11 +68,514 @@ transaction is confirmed are also not persistent - until you know the script
+    - the estimated weight of the tx is greater than 400,000 (`MAX_STANDARD_TX_WEIGHT`)
+
+#### Rationale
+
+To signal the conclusion of exchange of transaction inputs and outputs.
+
+Upon successful exchange of `tx_complete` messages, both nodes
+should construct the transaction and proceed to the next portion of the
+protocol. For channel establishment v2, exchanging commitment transactions.
+
+For the `minimum fee` calculation see [BOLT #3](03-transactions.md#calculating-fees-for-collaborative-transaction-construction).
+
+The maximum inputs and outputs are capped at 252. This effectively fixes
+the byte size of the input and output counts on the transaction to one (1).
+
+### The `tx_signatures` Message
+
+1. type: 71 (`tx_signatures`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`sha256`:`txid`]
+    * [`u16`:`num_witnesses`]
+    * [`num_witnesses*witness`:`witnesses`]
+
+1. subtype: `witness`
+2. data:
+    * [`u16`:`len`]
+    * [`len*byte`:`witness_data`]
+
+#### Requirements
+
+The sending node:
+  - if it has the lowest total satoshis contributed, as defined by total
+    `tx_add_input` values, or both peers have contributed equal amounts
+    but it has the lowest `node_id` (sorted lexicographically):
+    - MUST transmit their `tx_signatures` first
+  - MUST order the `witnesses` by the `serial_id` of the input they
+    correspond to
+  - `num_witnesses`s MUST equal the number of inputs they added
+  - MUST use the `SIGHASH_ALL` (0x01) flag on each signature
+
+The receiving node:
+  - MUST fail the negotiation if:
+    - the message contains an empty `witness`
+    - the number of `witnesses` does not equal the number of inputs
+      added by the sending node
+    - the `txid` does not match the txid of the transaction
+    - the `witnesses` are non-standard
+    - a signature uses a flag that is not `SIGHASH_ALL` (0x01)
+  - SHOULD apply the `witnesses` to the transaction and broadcast it
+  - MUST reply with their `tx_signatures` if not already transmitted
+
+#### Rationale
+
+A strict ordering is used to decide which peer sends `tx_signatures` first.
+This prevents deadlocks where each peer is waiting for the other peer to
+send `tx_signatures`, and enables multiparty tx collaboration.
+
+The `witness_data` is encoded as per bitcoin's wire protocol (a CompactSize number
+of elements, with each element a CompactSize length and that many bytes following).
+
+While the `minimum fee` is calculated and verified at `tx_complete` conclusion,
+it is possible for the fee for the exchanged witness data to be underpaid.
+It is the responsibility of the sending peer to correctly account for the
+required fee.
+
+### The `tx_init_rbf` Message
+
+This message initiates a replacement of the transaction after it's been
+completed.
+
+1. type: 72 (`tx_init_rbf`)
+2. data:
+   * [`channel_id`:`channel_id`]
+   * [`u32`:`locktime`]
+   * [`u32`:`feerate`]
+   * [`tx_init_rbf_tlvs`:`tlvs`]
+
+1. `tlv_stream`: `tx_init_rbf_tlvs`
```

Comment:
I missed that earlier, but I believe `tx_init_rbf` and `tx_ack_rbf` should also have the `require_confirmed_inputs` TLVs that `open_channel2` and `accept_channel2` have? We shouldn't implicitly inherit that from the previous pending transaction (especially in the future context of splicing), and even if you keep setting it exactly like you did in `open_channel2` and `accept_channel2`, it doesn't hurt to be explicit?

Reply:
Note that since this is an `even` TLV, we should decide before merging the spec, as this will be incompatible with current readers of the `tx_init_rbf` message. I think we must do it, so we should do it ASAP! Another option though could be to use an odd tlv tag when adding this TLV to `tx_init_rbf` and `splice_init`.